File: server/logic/kmeans.py
#!/usr/bin/env python3
import sys
import json
import numpy as np
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

def run(df, params=None):
    """
    Run KMeans clustering on the given DataFrame
    
    Args:
        df: pandas DataFrame or list of dictionaries
        params: dictionary of parameters (optional)
    
    Returns:
        Dictionary with clustering results
    """
    # Set default parameters
    if params is None:
        params = {}
    
    n_clusters = params.get('n_clusters', 3)
    
    # Convert input to DataFrame if it's not already
    if not isinstance(df, pd.DataFrame):
        if isinstance(df, list):
            df = pd.DataFrame(df)
        else:
            return {"error": "Input data must be a DataFrame or list of dictionaries"}
    
    # Get only numeric columns
    X = df.select_dtypes(include='number')
    
    if X.empty:
        return {"error": "No numeric data available for clustering"}
    
    # Extract features
    features = X.values
    
    # Run KMeans
    kmeans = KMeans(n_clusters=n_clusters, random_state=42)
    labels = kmeans.fit_predict(features)
    
    # Calculate silhouette score
    silhouette = 0
    if n_clusters > 1 and n_clusters < len(features):
        try:
            silhouette = silhouette_score(features, labels)
        except:
            silhouette = 0
    
    # Convert data for result
    data_list = X.to_dict('records')
    
    # Format results
    result = {
        "data": data_list,
        "labels": labels.tolist(),
        "centroids": kmeans.cluster_centers_.tolist(),
        "stats": {
            "inertia": float(kmeans.inertia_),
            "silhouette_score": float(silhouette),
            "iterations": int(kmeans.n_iter_)
        },
        "explanation": f"KMeans clustering with {n_clusters} clusters applied to {len(data_list)} data points with {X.shape[1]} features."
    }
    
    return result

def run_kmeans(data, n_clusters=3):
    """Legacy function for backward compatibility"""
    df = pd.DataFrame(data)
    return run(df, {'n_clusters': n_clusters})

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: python kmeans.py <data_file> [n_clusters]"}))
        sys.exit(1)
    
    try:
        # Read data file
        with open(sys.argv[1], 'r') as f:
            data = json.load(f)
        
        # Get number of clusters
        n_clusters = 3
        if len(sys.argv) > 2:
            n_clusters = int(sys.argv[2])
        
        # Convert to DataFrame
        df = pd.DataFrame(data)
        
        # Run KMeans
        result = run(df, {'n_clusters': n_clusters})
        
        # Print result as JSON
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

File: server/logic/kmeans.js
// KMeans implementation using simple-kmeans package
const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

// We'll use Python's scikit-learn for KMeans since it's more robust
function runKMeans(data, params = {}) {
  return new Promise((resolve, reject) => {
    // Default parameters
    const nClusters = params.nClusters || 3;
    
    // Create a temporary file to store the data
    const tempFile = path.join(__dirname, '../../temp_data.json');
    
    // Ensure we only use numeric data for clustering
    const numericData = data.map(row => {
      const numericRow = {};
      Object.keys(row).forEach(key => {
        if (typeof row[key] === 'number' && !isNaN(row[key])) {
          numericRow[key] = row[key];
        }
      });
      return numericRow;
    });
    
    // Write the data to a temporary file
    fs.writeFileSync(tempFile, JSON.stringify(numericData));
    
    // Path to Python script
    const pythonScript = path.join(__dirname, 'kmeans.py');
    
    // Create the Python script if it doesn't exist
    if (!fs.existsSync(pythonScript)) {
      const pythonCode = `
import sys
import json
import numpy as np
from sklearn.cluster import KMeans

# Load the data from the temporary file
with open(sys.argv[1], 'r') as f:
    data = json.load(f)

# Extract features (assuming all values are numeric)
features = []
for row in data:
    # Get all numeric values
    row_values = list(row.values())
    if row_values:  # Make sure we have values
        features.append(row_values)

# Convert to numpy array
X = np.array(features)

# Fit KMeans
n_clusters = int(sys.argv[2])
kmeans = KMeans(n_clusters=n_clusters, random_state=42)
kmeans.fit(X)

# Prepare results
result = {
    "charts": {
        "centroids": kmeans.cluster_centers_.tolist()
    },
    "stats": {
        "inertia": float(kmeans.inertia_),
        "n_iter": int(kmeans.n_iter_)
    },
    "tables": {
        "labels": kmeans.labels_.tolist()
    },
    "explanation": "KMeans clustered your data based on numeric features using Euclidean distance."
}

# Output as JSON
print(json.dumps(result))
      `;
      fs.writeFileSync(pythonScript, pythonCode);
    }
    
    // Run the Python script
    const pythonProcess = spawn('python', [pythonScript, tempFile, nClusters.toString()]);
    
    let resultData = '';
    let errorData = '';
    
    pythonProcess.stdout.on('data', (data) => {
      resultData += data.toString();
    });
    
    pythonProcess.stderr.on('data', (data) => {
      errorData += data.toString();
    });
    
    pythonProcess.on('close', (code) => {
      // Clean up the temporary file
      if (fs.existsSync(tempFile)) {
        fs.unlinkSync(tempFile);
      }
      
      if (code !== 0) {
        console.error('Python process error:', errorData);
        reject(new Error(`Python process exited with code ${code}: ${errorData}`));
        return;
      }
      
      try {
        const result = JSON.parse(resultData);
        resolve(result);
      } catch (error) {
        reject(new Error(`Failed to parse KMeans result: ${error.message}`));
      }
    });
  });
}

module.exports = {
  run: runKMeans
};